### PR TITLE
Loans: Add Runtime API

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

Fix partially #1174 for loans. I have not wanted to add the pool part because I'm not pretty involved in the index pool feature, and things have changed a bit from the last specs regarding this. 

Also, not sure what we gain by merging pool & loan things in the same RuntimeAPI call. The user can make two API requests, and we maintain things isolated. If making two calls comes with some performance cost (that, as a first view, doesn't, because both calls can be done concurrently), then I think we can add an API wrapper that performs both API requests, keeping things isolated.

## Changes and Descriptions
- [x] Add an RPC to get `ActiveLoan` data and computed values as `present_value` and `interest_accrued`. 
- [x] ~~Integration tests~~ => Can be done in another PR => #1458

### CODE REVIEW

File: src/rpc/loans.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	#[method(name = "loans_portfolioLoan")]
+	fn portfolio_loan(
+		&self,
+		pool_id: PoolId,
+		loan_id: LoanId,
+		at: Option<BlockHash>,
+	) -> RpcResult<ActiveLoanInfo<T>>;
+}
+
+pub struct Loans<C, P> {
```
Comment (cdamian) [inline src/rpc/loans.rs:31]: nit - rename second generic param to `Loans<C, Block>`?

**Review by cdamian — DISMISSED** [review]

File: pallets/loans/src/entities/loans.rs [hunk 2]
```diff
@@ -489,3 +491,47 @@ impl<T: Config> ActiveLoan<T> {
 		self.schedule.maturity = crate::types::Maturity::fixed(duration);
 	}
 }
+
+/// Data containing an active loan with extra computed.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebugNoBound, TypeInfo, MaxEncodedLen)]
+#[scale_info(skip_type_params(T))]
+#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "std", serde(bound = ""))]
+pub struct ActiveLoanInfo<T: Config> {
```
Comment (lemunozm) [inline pallets/loans/src/entities/loans.rs:500]: This is a bit tricky and probably useful for futures runtime APIs.
The line with `bound = ""` is mandatory for cases where you have both, the `scale_info` and `derive(serde)` lines. Serde defaults to adding `Serialize` and `Deserialize` bounds to `T`,  but T, in this case, does not need to be serializable, it's only used to fetch information from associated types. With `bounds = ""`, we remove those implicit bounds from serde.

**Comment by cdamian** [comment]
@lemunozm Can you also add an integration test for this similar to the one that we have for rewards?

**Comment by lemunozm** [comment]
Yeah, I wanted to do it, but looking into it, I've realized it would be quite time-consuming (creating a pool, creating assets, registering permissions, creating investors, investing in the pool, borrowing from loans to make the "active" loan, ...) just for being able to make the runtime API call, and right now I need to put efforts in order task for the release 😓.

Also, I would need to create integration tests for all loans related things. Right now, it's well-tested at pallet level but not at runtime level. Issue to track this work here: #1458

File: pallets/loans/src/entities/loans.rs (hunk 2 above)
Comment (wischli) [reply pallets/loans/src/entities/loans.rs:500]: Thanks for clarifying this. So this is the `serde` equivalent of `scale_info(skip_type_params(T)` more or less?

File: pallets/loans/Cargo.toml [hunk 3]
```diff
@@ -63,6 +64,7 @@ std = [
   "sp-io/std",
   "strum/std",
   "orml-traits/std",
+  "serde/std",
```
Comment (wischli) [inline pallets/loans/Cargo.toml:67]: Since it's optional, shouldn't we apply `?`
```suggestion
  "serde?/std",
```

File: libs/traits/Cargo.toml [hunk 4]
```diff
@@ -43,6 +44,7 @@ std = [
   "sp-std/std",
   "cfg-primitives/std",
   "scale-info/std",
+  "serde/std",
```
Comment (wischli) [inline libs/traits/Cargo.toml:47]: Since it's optional, shouldn't we apply `?`
```suggestion
  "serde?/std",
```

**Review by wischli — COMMENTED** [review]
LGTM apart from my minor nit and the supposedly missing `#[serde(rename_all = "camelCase")]` for all call input arguments based on [this comment](https://github.com/polkadot-js/api/issues/5529#issuecomment-1465271895).

File: pallets/loans/Cargo.toml (hunk 3 above)
Comment (lemunozm) [reply pallets/loans/Cargo.toml:67]: Good one!

File: pallets/loans/src/entities/loans.rs (hunk 2 above)
Comment (lemunozm) [reply pallets/loans/src/entities/loans.rs:500]: Yeah! More or less. Also, if there were other bounds, and you do not want to "reset them" you should add `bound = "K: MyOtherBound"`. That way you've reset the `Serialize` bound to `T` and `K` but not the `MyOtherBound` to `K`.

File: libs/traits/Cargo.toml (hunk 4 above)
Comment (wischli) [reply libs/traits/Cargo.toml]: Wondering why adding this broke tests?

Comment (lemunozm) [reply libs/traits/Cargo.toml:47]: I think:
- `serde?/std` adds `serde/std` in case `serde` is enable
- `serde/srd` enables `serde` and adds `serde/std`

**Comment by lemunozm** [comment]
Much much simpler PR after removing the annoying RPC part. Thanks @wischli!

**Review by wischli — APPROVED** [review]
All hail the runtime API!

**Pull request merged**